workflow-core/src/main/java/com/squareup/workflow1/BaseRenderContext.kt

```diff
@@ -310,6 +310,37 @@ inline fun <T, reified W : Worker<T>, PropsT, StateT, OutputT>
   runningWorker(worker, typeOf<W>(), key, handler)
 }
 
+/**
+ * Ensures the worker produce by [workerProvider] is running. When the [Worker] emits an output,
+ * [handler] is called to determine the [WorkflowAction] to take. When the worker finishes, nothing
+ * happens (although another render pass may be triggered).
+ *
+ * Unlike the other [runningWorker] methods, this one requires a mandatory [key] which is used to
+ * determine if workers are to be started, kept alive, or destroyed across multiple render passes.
+ * This allows the runtime to delay construction of workers until necessary and avoid reflection.
+ *
+ * @param key A string key that is used to distinguish [Worker]s emitted by [workerProvider].
+ */
+/* ktlint-disable parameter-list-wrapping */
+inline fun <T, reified W : Worker<T>, PropsT, StateT, OutputT>
+    BaseRenderContext<PropsT, StateT, OutputT>.runningWorker(
+  noinline workerProvider: () -> W,
+  key: String,
+  noinline handler: (T) -> WorkflowAction<PropsT, StateT, OutputT>
+) {
+/* ktlint-enable parameter-list-wrapping */
+  runningWorker(LazyWorker(key, workerProvider), LazyWorkerType, key, handler)
```

Comment:
@zach-klippenstein the `typeOf` has been replaced with a singleton `LazyWorkerType` used across all invocations that pass in a key, assuming these keys will be unique.

Referring to:

>And the implementation of RenderContext can't run at all to do this sort of logic until the type has been captured anyway

Would `LazyWorkerType` break RenderContext implementations in some insidious way? If so, I'm guessing I would have to implement a `String` key-based `WorkflowIdentifier` for my usecase, right?

Reply:
It should work, and if you have a use case for it, this technique should be fine. 

However changing the use of type information in the standard worker builder functions at this point could break code that is written assuming they take both type and key into account.

It would be safer to create separate builder functions that are explicit about ignoring type (eg `asUntypedWorker()`). And for that matter, I think `UntypedWorker` would be a better name for this construct, since the most significant thing about it is its lack of type information, not its laziness (all workers are "lazy" in a sense).

I think I would be ok with landing this stuff if it didn't change any existing behavior. @rjrjr WDYT?

Reply:
My kneejerk is always going to be to push back against additions to the library that users can implement themselves. Do you both think its of general interest?

Reply:
I agree, although technically Worker itself falls under this category now I believe.

I'd rather not merge this in the core library for now, keep the API surface smaller.

Reply:
Makes sense, I can add the builders to our project instead.